lib/convict.js

```diff
@@ -171,6 +171,17 @@ function validate(instance, schema, strictValidation) {
     }
     delete flatInstance[name];
 
+    // ignore nested keys of 'object' config params 
+    if (schemaItem.format === 'object') {
+      const startsWithName = new RegExp('^' + name);
```

Comment:
Instead of using a regular expression, you should probably just use [`String.startsWith()`][startswith]. Consider the following (contrived) example:
```javascript
const convict = require('.');

const schema = {
  parent: {
    child: {
      default: {},
      format: Object
    }
  }
};

const config = convict(schema);

const data = {
  parent: {
    child: { five: 5 }
  },
  parent_child: 5
};

config.load(data);

config.validate({ allowed: 'strict' });
```
In this situation, `flatInstance` looks like this:
```json
{
  "parent_child": 5,
  "parent.child.five": 5
}
```
And `flatSchema` looks like this*:
```json
{
  "parent.child": {
    "default": { "five": 5 },
    "format": "object",
    "_format": Function
  }
}
```
The regular expression being matched will be `^parent.child`, which will match `parent_child` and subsequently remove it from later consideration. Other issues will come into play if the name contains other special regex characters (e.g. an object-type field named `$key`). By switching to `String.startsWith()`, you avoid that issue.

Additionally, instead of matching on `name` directly, you should match on `${name}.`. To demonstrate why, consider this case:
```javascript
const convict = require('.');

const schema = {
  obj: {
    default: {},
    format: Object
  }
};

const config = convict(schema);

const data = {
  obj: { five: 5 },
  obj2: 5
};

config.load(data);

config.validate({ allowed: 'strict' });
```
In this situation, `flatInstance` looks like this:
```json
{
  "obj2": 5,
  "obj.five": 5
}
```
And `flatSchema` looks like this:
```json
{
  "obj": {
    "default": { "five": 5 },
    "format": "object",
    "_format": Function
  }
}
```
In this situation, if we match on `name` directly (either via regular expression or `String.startsWith()`), `obj` compared to the beginning of both `obj.five` and `obj2`, and both succeed. However, we only want to discard `obj.five`. If we instead match on `${name}.` (i.e. `obj.`), only `obj.five` is discarded from later consideration.

It's probably a good idea to add these cases to the test suite.

\* As an aside, this investigation appears to have unearthed a bug with defaults in this sort of situation. The default of `parent.child` should be `{}`, but here is overridden into `{five: 5}`. I'll see if I can track that down.

[startswith]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/startsWith

Reply:
Other than this, the change looks good to me. Add those tests and switch to `String.startsWith()` and we're good.

Reply:
ouch, sorry for missing this review, and thanks! I avoided `startsWith` in order to make this a non-breaking change requiring Node version >=6.4, but I'm happy to make the change and bump your [package.json `engines` value](https://github.com/mozilla/node-convict/blob/master/package.json#L21-L23)

Reply:
In that case, to avoid changing the Node version required, go with `key.lastIndexOf(name + '.', 0) === 0`.